### ISSUE TITLE
Multiple bindings for initBindings/VALUES?

### ISSUE THREAD

**Issue by stefanbischof** [RDFLib/rdflib#2012]
Is it possible to define multiple bindings for the `initBindings` parameter (SPARQL `VALUES`)?

How would you encode this example from the SPARQL 1.1 spec with `initBindings` (while I don't need `UNDEF` yet)?

```sparql
VALUES (?x ?y) {
  (:uri1 1)
  (:uri2 UNDEF)
}
```

`initBindings` is a dictionary which allows only RDF terms, right? Would it make sense to allow a list of RDF terms there?

**Comment by aucampia** [comment]
@stefanbischof some code would be helpful here, I'm actually not entirely sure what you are asking, and what the context is, are you talking about query? initBindings is a dictionary and you could just add multiple keys with different values, though if you could give an example with code of what you want it would be much simpler to understand the query.

**Comment by stefanbischof** [comment]
> initBindings is a dictionary and you could just add multiple keys with different values

yes, the problem is, you can only give one single value to each variable, while SPARQL allows to specify several "rows" of values not only single values. Another example:
```sparql
VALUES (?x ?y) {
  (2 1)
  (3 5)
}
```
Overall I'd like to specify initBindings somehow like this. There is several ways how to encode this in python. One way would be to allow lists of values. Corresponding example:

```python
initBindings={"x": [2,3], "y": [1,5]}
```

**Comment by gjhiggins** [comment]
FFR: https://www.w3.org/TR/sparql11-query/#inline-data pertains.

**Comment by aucampia** [comment]
> Overall I'd like to specify initBindings somehow like this.

And this would be an argument to `Graph.query`?

I think I understand but if you give complete SPARQL query, a complete graph, and code with expected outcome it would be easier to ensure that we assess this request correctly and that if someone eventually looks at this that they actually do implement what you want.

**Comment by stefanbischof** [comment]
after using `initBindings` a bit more: does it **replace the variables** in the query or does it **append a `VALUES` clause** to the query? At first I assumed the `VALUES` case, but now I got the impression that variables are actually replaced.

**Comment by sivy** [comment]
@stefanbischof As far as I can determine (from a few days spent in PDB) in `query()`, `initBindings` behaves like SQL parameters - the variables are replaced in the query.

```python
query = """
SELECT ?name
WHERE {
    ?actor a as:Person .
    ?actor as:name ?name .
    ?actor as:preferredUsername ?account .
}
"""
user_query = sparql.prepareQuery(query, initNs=NAMESPACES)
users = graph.query(
    user_query,
    initBindings={
        "account": rdflib.Literal("someusername"),
    },
)
```

The query after `initBindings` are used will be:

```sparql
SELECT ?name
WHERE {
    ?actor a as:Person .
    ?actor as:name ?name .
    ?actor as:preferredUsername 'someusername' .
}
```

**Comment by sanmai-NL** [comment]
@stefanbischof Can this be closed? Seems like an unclear feature, perhaps some docs would help, but not a work item otherwise?

**Comment by stefanbischof** [comment]
Yes, you can close it. I originally thought the initbindings would behave like a sparql values clause, when it is only variable substitution.

**Comment by sanmai-NL** [comment]
@stefanbischof Understood, but I'm not a maintainer here. Just asking for follow up. Could you close it yourself, please?
